misk-crypto/src/main/kotlin/misk/crypto/CiphertextFormat.kt

```diff
@@ -0,0 +1,317 @@
+ *     )*
+ *   ]
+ * ]
+ * ```
+ *
+ * The final output will be serialized using the following format:
+ * ```
+ * [ 0xEE: magic+version ]
+ * [ varint: AAD length ]
+ * [ AAD ]
+ * [ tink ciphertext ]
+ * ```
+ *
+ * For the full documentation of the [CiphertextFormat] serialization, read FORMAT.md
+ */
+class CiphertextFormat private constructor() {
+
+  companion object {
+    /**
+     * Current version of the encryption packet schema
+     */
+    const val CURRENT_VERSION = 0xEE
+
+    /**
+     * Serializes the given [ciphertext] and associated encryption context to a [ByteArray]
+     */
+    fun serialize(ciphertext: ByteArray, aad: ByteArray?): ByteArray {
+      val outputStream = ByteStreams.newDataOutput()
+      outputStream.writeByte(CURRENT_VERSION)
+      if (aad == null) {
+        outputStream.write(byteArrayOf(0))
+      } else {
+        outputStream.write(encodeVarInt(aad.size))
+        outputStream.write(aad)
+      }
+      outputStream.write(ciphertext)
+      return outputStream.toByteArray()
+    }
+
+    /**
+     * Extract the ciphertext and associated authentication data from the [serialized] ByteArray.
+     *
+     * This method also compares the given [context] to the serialized AAD
+     * and will throw an exception if they do not match.
+     */
+    fun deserialize(serialized: ByteArray, context: Map<String, String>?): Pair<ByteArray, ByteArray?> {
+      val src = DataInputStream(ByteArrayInputStream(serialized))
+      val version = src.readByte()
+      if (version != CURRENT_VERSION.toByte()) {
+        throw InvalidCiphertextFormatException("invalid version: $version")
+      }
+      val aad: ByteArray?
+      val ecSize = decodeVarInt(src)
+      if (ecSize > 0) {
+        aad = ByteArray(ecSize)
+        src.readFully(aad)
+      } else {
+        aad = null
+      }
+
+      val serializedEncryptionContext = serializeEncryptionContext(context)
+      if ((aad == null && serializedEncryptionContext != null) ||
+          (aad != null && serializedEncryptionContext == null)) {
+        throw InvalidEncryptionContextException("encryption context doesn't match")
+      }
+      if (aad != null) {
+        val equals = serializedEncryptionContext?.contentEquals(aad)
+            ?: throw InvalidEncryptionContextException("encryption context doesn't match")
+        if (!equals) {
+          throw InvalidEncryptionContextException("encryption context doesn't match")
+        }
+      }
+      val ciphertext = readCiphertext(src)
+      return Pair(ciphertext, aad)
+    }
+
+    /**
+     * Serializes the encryption context to a [ByteArray] so it could be passed to Tink's
+     * encryption/decryption methods.
+     */
+    fun serializeEncryptionContext(context: Map<String, String>?): ByteArray? {
+      if (context == null || context.isEmpty()) {
+        return null
+      }
+
+      val regex = Regex("[-_\\s]")
+      val filteredKeySet = context.keys.map {
```

Comment:
This is user supplied data, and we (@yan and I) introduced this idea because we wanted to make sure encryption context keys are as clear as possible.
In general we allow pretty much any string as a key/value in this map, but if you supply something like `{"some_key": "value1", "Some-Key": "another value"}` it's just calling for trouble and ambiguity.

I think it's helpful because it forces the user to choose better, more descriptive names.
But, it's also a runtime error which means we can fail encrypting something that you don't have control over which sucks.

Reply:
I've decided to get rid of this check for a couple of reasons:
- This is part of an encryption library, there's absolutely no reason to fail here because we don't like how the encryption context looks like
- The encryption context could be pretty complex, and may be generated by a few different components that might add similar looking key names
- Having this also means that if you created a key at some point by mistake, e.g. `"some key"` and you wanted to change it to `"some-key"`, it'd be very hard and annoying to do